### ISSUE TITLE
Add a mechanism to stage bigquery and cloud storage partitions until complete

### ISSUE THREAD

**Issue by relud** [mozilla/gcp-ingestion#121]
Problem: backfilling data after gcp ingestion downtime requires updating any derived data sets that may have read the partition before the backfill, including scheduled queries from re:dash.

Solution: for non-streaming tables (in both bigquery and cloud storage), require that partitions be staged then promoted atomically (or as close to atomic as possible), and document suspending promotion.

Implementation: Deliver ingestion data to a staging location and use airflow to schedule operations that promote completed partitions. Before promoting data the operation should attempt to validate that data is complete within some threshold. Make sure that both data ops and data platform engineers are trained in suspending promotion in the case of potential incomplete data, such as from downtime.

**Comment by relud** [comment]
The goal here is that once a partition can be read it is complete and never needs to be read again.

This provides consumers a simple contract for ensuring that the data they see is complete.

**Comment by relud** [comment]
In the case that an existing partition must be updated or backfilled we should have a documented process for updating data platform managed tables and communicating the need to backfill to consumers.

**Comment by jklukas** [comment]
Do we know how expensive it is to promote data? For BQ case, are we just renaming a table? For GCS step, I assume we actually need to copy and delete?

We can probably write the Airflow DAGs in such a way that we aren't introducing significant latency (we do the validity check and release the data immediately after the job that creates them finishes).

I do like the extra guarantees this process gives for clients. The biggest unknown to me is how to implement the data validity check and whether that needs to be something we maintain for each separate dataset.

**Comment by relud** [comment]
I think the data validity check should be generic and look for a configurable maximum gap in submission timestamps

**Comment by whd** [comment]
My initial thought is that this sounds reasonable to implement for data gaps (a specific kind of ingestion error), but sounds like it would be more difficult to handle for e.g. data validation errors such as when we change a schema erroneously or the data itself changes. In my experience on the current pipeline we've had to backfill datasets for missing data once, and for schema issues more often than that. I think in general the data validity check is the difficult piece and that one as simple as "configurable maximum gap" isn't going to be sufficient for all cases.

**Comment by relud** [comment]
> Do we know how expensive it is to promote data?

I think it should be relatively cheap, but I haven't attempted a cost analysis

> For BQ case, are we just renaming a table?

I would expect something like
```bash
bq cp staging_dataset.table$day dataset.table$day
bq query "
    DELETE staging_dataset.table$day
    WHERE document_id NOT IN (
        SELECT document_id FROM dataset.table$day
    )
"
```

> For GCS step, I assume we actually need to copy and delete?

I would expect something like:
```
gsutil -m mv gs://bucket/staging/table/day gs://bucket/table/day
gsutil cp /dev/null gs://bucket/table/day.complete
```
 and maybe something to add the partition to HIVE.

**Comment by relud** [comment]
@whd good point. I can think of two things immediately for validating other cases in a generic way:

1. compare error volume to total data volume for both the individual table and the relevant decoder, and refuse if the error ratio is above some configurable threshold
1. attempt to detect anomalous data volume for the partition
   - pretty sure this option is unreasonably error prone

**Comment by whd** [comment]
(1) can cap the expected error level for which we would need to backfill (what is "good enough"), but it may be difficult to detect small volume changes e.g. a new measure is added to nightly and rides the trains, but due to low volume we don't notice the schema is wrong until it hits beta (and is only wrong on a specific operating system in a specific configuration).

The core problem is that the contract "this data is complete and correct" is somewhat difficult to enforce generally in the face of any change to the system. What (say) in the case of such an accidental type change to an important metric that requires special re-processing to forcibly coerce the type? It's a business decision to decide how and if we should reprocess the data in this case, which can't necessarily be handled generically.

That it is difficult to enforce the contract is not a reason not to attempt to do it, but I would be wary of making this a SLO without more discussion and policy decisions.

**Comment by relud** [comment]
@whd excellent points, I think that we should reduce the scope of this.

I think the scope and contract here should only be "the data is complete within a margin of error".

The correctness of data would be out of scope, unless data is determined to be incorrect with sufficient frequency to make "complete" useless.

The language used to record this decision would need to be very clear about what constitutes complete.

I don't think anything should be an SLO unless we can measure it and define specific consequences for missing it (I'm a big fan of "suspend all feature work and spend <amount of time> working only on stability work that helps meet the SLO")

> we don't notice the schema is wrong until it hits beta

> requires special re-processing to forcibly coerce the type

I think incorrect schemas should be specifically called out as a non-downtime case that requires backfill, and backfilling should be documented process (and automated where possible).

**Comment by jklukas** [comment]
I believe the current system we have for populating historical ping tables from live ping tables mostly satisfies the intention of this issue.

**Comment by relud** [comment]
i agree
